servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/AbstractTimeoutHttpFilter.java

```diff
@@ -151,6 +158,64 @@ private Executor contextExecutor(final HttpRequestMetaData requestMetaData,
                 context.executor() : context.ioExecutor();
     }
 
+    // This is responsible for ensuring that we don't abandon the response resources due to the use of Single.timeout.
+    // It does so by retaining a reference to the StreamingHttpResponse in case we receive a cancellation. It will
+    // then attempt to drain the resource. We're protected from multiple subscribes because the TimeoutSingle will
+    // short circuit also on upstream cancellation, although that is not obviously correct behavior.
+    private static final class CleanupSubscriber implements SingleSource.Subscriber<StreamingHttpResponse> {
+
+        private static final AtomicReferenceFieldUpdater<CleanupSubscriber, Object> stateUpdater =
+                AtomicReferenceFieldUpdater.newUpdater(CleanupSubscriber.class, Object.class, "state");
+
+        private static final Object COMPLETE = "cancelled";
+
+        private final SingleSource.Subscriber<? super StreamingHttpResponse> delegate;
+        @SuppressWarnings("unused")
+        @Nullable
+        private volatile Object state;
+
+        CleanupSubscriber(final SingleSource.Subscriber<? super StreamingHttpResponse> delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void onSubscribe(Cancellable cancellable) {
+            delegate.onSubscribe(() -> {
+                onCancel();
+                cancellable.cancel();
+            });
+        }
+
+        @Override
+        public void onSuccess(@Nullable StreamingHttpResponse result) {
+            assert result != null;
+            if (stateUpdater.compareAndSet(this, null, result)) {
+                // We win.
+                delegate.onSuccess(result);
+            } else {
+                // we lost to cancellation. No need to send it forward.
+                clean(result);
+            }
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            assert !(stateUpdater.get(this) instanceof StreamingHttpResponse);
```

Comment:
nit: for "get" we typically access the `state` directly, no need to go through `stateUpdater`

Reply:
I'll change it for consistency, but iirc using the updater is a more generally correct pattern. Implementations of atomic updaters can vary depending on the type, platform, and/or JVM implementation. See `AtomicLongFieldUpdater.LockedUpdater` for an example.
I realize this case is a pointer, but what I'm suggesting is that it's never incorrect to use the Updater methods whereas there is more nuance to consider when using the raw field so consistently using the Updater is an easier pattern to get right.